processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java

```diff
@@ -46,18 +47,26 @@ static int compare(Row row, String aggregationName, Number value, Map<String, Ag
         metricValueObj = aggregators.get(aggregationName).finalizeComputation(metricValueObj);
       }
 
-      if (metricValueObj instanceof Long) {
-        long l = ((Long) metricValueObj).longValue();
-        return Longs.compare(l, value.longValue());
+      if (metricValueObj instanceof Long || metricValueObj instanceof Integer) {
+        // Cast ints metrics to longs. It makes the code simpler and won't change the result.
+        long n = ((Number) metricValueObj).longValue();
+
+        if (value instanceof Long || value instanceof Integer) {
+          return Longs.compare(n, value.longValue());
+        } else if (value instanceof Double || value instanceof Float) {
+          // Use BigDecimal for comparison, a convenient way to handle edge cases without worrying about them
+          // ourselves. Cast the value to double since that's what BigDecimal wants. Should be fine since doubles can
+          // represent all float values.
+          return BigDecimal.valueOf(n).compareTo(BigDecimal.valueOf(value.doubleValue()));
+        } else {
+          throw new ISE("Number was[%s]?!?", value.getClass().getName());
+        }
       } else if (metricValueObj instanceof Float) {
-        float l = ((Float) metricValueObj).floatValue();
-        return Floats.compare(l, value.floatValue());
+        float n = (Float) metricValueObj;
+        return Floats.compare(n, value.floatValue());
```

Comment:
can we combine float and double also similar to long and int... also check type of value instance and use BigDecimal if it is not double? ... to keep it similar to long/int case.

Reply:
@himanshug I edited the code to simplify it, and factored out the BigDecimal stuff to a helper method. Now the branches are more similar.

Reply:
thanks